src/stacked_borrows/diagnostics.rs

```diff
@@ -1,98 +1,286 @@
 use smallvec::SmallVec;
+use std::fmt;
 
-use rustc_middle::mir::interpret::{AllocId, AllocRange};
+use rustc_middle::mir::interpret::{alloc_range, AllocId, AllocRange};
+use rustc_middle::mir::RetagKind;
 use rustc_span::{Span, SpanData};
 use rustc_target::abi::Size;
 
 use crate::helpers::CurrentSpan;
-use crate::stacked_borrows::{err_sb_ub, AccessKind};
+use crate::stacked_borrows::{err_sb_ub, AccessKind, Permission};
 use crate::*;
 
 use rustc_middle::mir::interpret::InterpError;
 
 #[derive(Clone, Debug)]
 pub struct AllocHistory {
-    // The time tags can be compressed down to one bit per event, by just storing a Vec<u8>
-    // where each bit is set to indicate if the event was a creation or a retag
-    current_time: usize,
-    creations: smallvec::SmallVec<[Event; 2]>,
-    invalidations: smallvec::SmallVec<[Event; 1]>,
+    id: AllocId,
+    creations: smallvec::SmallVec<[Creation; 1]>,
+    invalidations: smallvec::SmallVec<[Invalidation; 1]>,
     protectors: smallvec::SmallVec<[Protection; 1]>,
+    current_operation: Operation,
+    current_offset: Size,
```

Comment:
So the core approach seems to be to not pass these things down the call stack, but to implicitly store them in global state.

My concern here is things getting our of sync. How do you ensure that these fields are always up-to-date? It seems very easy to forget to update them and then things will start to behave rather strangely. That's why I prefer prefer passing information down explicitly as function arguments.

To make things worse, some information is still passed via arguments (`provoking_access`), so the flow of diagnostic information is not even consistent.

So I wonder if it would make sense to a type like `Operation` or `DiagnosticCx` or whatever that accumulates the necessary information, but that is threaded through explicitly. We could even have the type reflect, e.g., whether an offset has already been set or not, by having `Operation` and `OperationAtOffset`. Since this is all stack-like, only an `&Operation` ever needs to actually be passed down, so this should also be rather efficient in terms of how much data is copied. But crucially, we can never accidentally be using information from the *previous* operation.

Having this in `AllocExtra` makes me uneasy.

Reply:
I tried bundling everything related to diagnostics into a `DiagnosticCx` and I think overall the design is a lot simpler inside the main stacked borrows implementation. I think I just skimmed your comment here before implementing something, so I'm happy to adjust closer to what you describe if you don't like what I came up with.